### DISCUSSION TITLE
How to prevent dokkatoo from running on `gradlew build`

### DISCUSSION THREAD

**Discussion by odenix** [adamko-dev/dokkatoo#203] [Q&A] [answered]
I don't want dokkatoo to run when I type `./gradlew build`, which is the default command run by developers and on-push CI builds.
How do I achieve this?

Dokkatoo version: 2.2.0

**Comment by aSemy** [comment]
Hi, I can understand why you wouldn't want Dokkatoo to run every time. However, I'm not sure how that is happening in your build. By default, Dokkatoo doesn't add any task links to the standard lifecycle tasks.

Can you run `./gradlew build --scan` to create a build scan, and investigate why the Dokkatoo generation tasks are running?

**Reply by odenix** [reply]
Here's what the task-tree plugin prints:

```
:build
+--- :assemble
|    +--- :dokkatooGenerateModuleHtml
```

This is a tiny Java (not Kotlin) project built with Gradle 8.7 that I'm preparing to open-source.
In the `plugins` section I have:

```
plugins {
  `java-library`
  alias(libs.plugins.dokkaHtml) // dokkaHtml = { id = "dev.adamko.dokkatoo-html", version = "2.2.0" }
}
  ```
  
I also have:
```
dependencies {
  dokkatooPluginHtml(libs.dokkaJava) // dokkaJava = { module = "org.jetbrains.dokka:kotlin-as-java-plugin", version = "1.9.20" }
}

dokkatoo {
  pluginsConfiguration.html {
    footerMessage.set("foobar")
  }
  dokkatooSourceSets {
    register("main") {
      jdkVersion.set(17) 
      documentedVisibilities(VisibilityModifier.PUBLIC, VisibilityModifier.PROTECTED)
      sourceRoots = fileTree("src/main/java/org/translatenix/foobar") {
        include("*.java")
      }
    }
  }
}
```

By the way, I had to add this workaround for dokkatoo:
```
configurations.all {
  // failOnVersionConflict() gives the  following conflicts for dokkatoo:
  // > Conflicts found for the following modules:
  //   - org.jetbrains.kotlin:kotlin-reflect between versions 1.9.22 and 1.4.21
  //   - org.jetbrains.kotlin:kotlin-stdlib between versions 1.9.22, 1.8.20 and 1.9.0
  //   - com.fasterxml.jackson.core:jackson-databind between versions 2.12.7.1 and 2.12.7
  //   - org.jetbrains.kotlin:kotlin-stdlib-jdk8 between versions 1.9.0, 1.8.0 and 1.8.20
  //   - org.jetbrains.kotlin:kotlin-stdlib-common between versions 1.9.22, 1.9.0 and 1.8.20
  //   - org.jetbrains:annotations between versions 23.0.0 and 13.0
  //   - org.jetbrains.kotlin:kotlin-stdlib-jdk7 between versions 1.9.0 and 1.8.0
  if (!name.contains("dokka")) { 
    resolutionStrategy {
      failOnVersionConflict()
      failOnNonReproducibleResolution()
    }
  }
}
```

**Reply by aSemy** [reply]
Ahh, I've found out why. 

For some mysterious reason, when a Configuration is set to 'visible', Gradle will always run the tasks that populate that Configuration. And Dokkatoo does just that:

https://github.com/adamko-dev/dokkatoo/blob/8fa0d527ff075a85ef5365a7d11de88d6ffd2b73/modules/dokkatoo-plugin/src/main/kotlin/internal/gradleUtils.kt#L53-L55


This is really surprising behaviour by Gradle. It's not documented at all, and impossible to intuit.

I'll make a fix, which will be out in the next version. 

For now, you can try overriding the default 'isVisible' for Dokkatoo configurations:


```kotlin
// build.gradle.kts

configurations
  .matching { "dokkatoo" in it.name.lowercase() }
  .configureEach {
    isVisible = false
  }
```

**Reply by odenix** [reply]
Thanks for help. Unfortunately it doesn't make a difference.

**Reply by odenix** [reply]
Does every build that uses dokkatoo suffer from this problem?

**Reply by aSemy** [reply]
Hmm, strange that the the workaround didn't work. But thanks for trying!

Yes, all builds that trigger the 'assemble' task will also trigger Dokkatoo tasks, although the impact will be limited due to incremental build, meaning that the Dokkatoo tasks won't run unless the input files have changed.

**Reply by odenix** [reply, marked as answer]
I'm glad to confirm that 2.3.0 fixes this problem.

**Reply by aSemy** [reply]
Great! Thank you very much for noticing and asking about it.
